// Run the following command in the root directory to re-generate the .py and .cs file
// protoc -I="Shared" --python_out="Client" --mypy_out="Client" --csharp_out="Server/Packets" "packets.proto"

syntax = "proto3";
package packets;
option csharp_namespace = "Moonlapse.Server.Packets";

import "google/protobuf/descriptor.proto";
extend google.protobuf.FieldOptions {
    bool encrypted = 9999;
}

message OkPacket {
    string message = 1;
}

message DenyPacket {
    string reason = 1;
}

message LoginPacket {
    string username = 1;
    string password = 2;
}

message RegisterPacket {
    string username = 1;
    string password = 2;
}

message ChatPacket {
    string name = 1;
    string message = 2;
}

message PublicRSAKeyPacket {
	bytes key = 1;
}

message AESKeyPacket {
    bytes key = 1;
}

message Packet {
    oneof type {
        OkPacket ok = 1 [(encrypted) = false];
        DenyPacket deny = 2 [(encrypted) = false];
        LoginPacket login = 3 [(encrypted) = true];
        RegisterPacket register = 4 [(encrypted) = true];
        ChatPacket chat = 5 [(encrypted) = false];
        PublicRSAKeyPacket public_rsa_key = 6 [(encrypted) = false];
        AESKeyPacket aes_key = 7 [(encrypted) = true];
    }
}